src/dso_api/dynamic_api/temporal.py

```diff
@@ -0,0 +1,144 @@
+        """Parse and validate the received date"""
+        if value == "*":
+            return "*"
+
+        try:
+            if "T" in value or " " in value:
+                return make_aware(datetime.fromisoformat(value))
+            else:
+                return date.fromisoformat(value)
+        except ValueError:
+            raise ValidationError(
+                f"Invalid date or date-time format for '{dimension}' parameter!"
+            ) from None
+
+    def filter_object_version(self, queryset: models.QuerySet) -> models.QuerySet:
+        """Apply an object-level filter such as "?volgnummer=...".
+        This should only be called for the top-level object.
+        """
+        if queryset.model.table_schema() != self.table_schema:
+            raise ValueError("QuerySet model type does not match")
+
+        if not self.is_versioned:
+            return queryset
+        elif self.version_value is None:
+            return self.filter_queryset(queryset)
+        else:
+            return queryset.filter(**{self.version_field: self.version_value})
+
+    def filter_queryset(self, queryset: models.QuerySet) -> models.QuerySet:
+        if queryset.model.table_schema() != self.table_schema:
+            raise ValueError("QuerySet model type does not match")
+
+        if not self.is_versioned or self.slice_value == "*":
+            # allow this method to be called unconditionally on objects,
+            # and allow ?geldigOp=* to return all data
+            return queryset
+
+        # Either take a given ?geldigOp=yyyy-mm-dd OR ?geldigOp=NOW()
+        slice_value = self.slice_value or now()
+        range_fields = self.slice_range_fields or self.default_range_fields
+
+        if range_fields is not None:
+            # start <= value AND (end IS NULL OR value < end)
+            start, end = map(to_snake_case, range_fields)
+            return queryset.filter(
+                Q(**{f"{start}__lte": slice_value})
+                & (Q(**{f"{end}__isnull": True}) | Q(**{f"{end}__gt": slice_value}))
+            ).order_by(f"-{start}")
+        else:
+            # Last attempt to get only the current temporal record; order by sequence.
+            # does SELECT DISTINCT ON(identifier) ... ORDER BY identifier, sequence DESC
+            # Alternative would be something like `HAVING sequence = MAX(SELECT sequence FROM ..)`
+            identifier = self.table_schema.identifier[0]
+            sequence_name = self.table_schema.temporal.identifier
+            return queryset.distinct(identifier).order_by(identifier, f"-{sequence_name}")
+
+    @property
+    def default_range_fields(self) -> Optional[TemporalDimensionFields]:
+        """Tell what the default fields would be to filter temporal objects."""
+        dimensions = self.table_schema.temporal.dimensions
+        return dimensions.get("geldigOp") or first(dimensions.values(), None)
+
+    @property
+    def url_parameters(self) -> dict[str, str]:
```

Comment:
I don't think this return type is correct. You might be returning a `self.slice_value` as the value in a dict. That instance variable is elsewhere defined to be: `slice_value: Optional[Union[Literal["*"], date, datetime]]. Hence the return type of this function should at least be: `-> dict[str, Union[str, date, datetime]]`

I am kind of amazed that `mypy` didn't caught this. Did you run it?
`

Reply:
Good catch! It should format the self.slice_value back to a string.